projectq/cengines/_swapandcnotflipper.py

```diff
@@ -0,0 +1,159 @@
+to the given connectivity graph. It also translates Swap gates to CNOTs if
+necessary.
+"""
+from copy import deepcopy
+
+from projectq.cengines import (BasicEngine,
+                               ForwarderEngine,
+                               CommandModifier)
+from projectq.meta import get_control_count
+from projectq.ops import (All,
+                          NOT,
+                          CNOT,
+                          H,
+                          Swap)
+
+
+class SwapAndCNOTFlipper(BasicEngine):
+    """
+    Flips CNOTs and translates Swaps to CNOTs where necessary.
+
+    Warning:
+        This engine assumes that CNOT and Hadamard gates are supported by
+        the following engines.
+
+    Warning:
+        This engine cannot be used as a backend.
+    """
+
+    def __init__(self, connectivity):
+        """
+        Initialize the engine.
+
+        Args:
+            connectivity (set): Set of tuples (c, t) where if (c, t) is an
+                element of the set means that a CNOT can be performed between
+                the physical ids (c, t) with c being the control and t being
+                the target qubit.
+        """
+        BasicEngine.__init__(self)
+        self.connectivity = connectivity
+
+    def is_available(self, cmd):
+        """
+        Check if the IBM backend can perform the Command cmd and return True
+        if so.
+
+        Args:
+            cmd (Command): The command to check
+        """
+        return self._is_swap(cmd) or self.next_engine.is_available(cmd)
+
+    def _is_cnot(self, cmd):
+        """
+        Check if the command corresponds to a CNOT (controlled NOT gate).
+
+        Args:
+            cmd (Command): Command to check
+        """
+        return (isinstance(cmd.gate, NOT.__class__) and
+                get_control_count(cmd) == 1)
+
+    def _is_swap(self, cmd):
+        """
+        Check if the command corresponds to a Swap gate.
+
+        Args:
+            cmd (Command): Command to check
+        """
+        return (get_control_count(cmd) == 0 and cmd.gate == Swap)
+
+    def _needs_flipping(self, cmd):
+        """
+        Return True if cmd is a CNOT which needs to be flipped around.
+
+        Args:
+            cmd (Command): Command to check
+        """
+        if not self._is_cnot(cmd):
+            return False
+
+        target = cmd.qubits[0][0].id
+        control = cmd.control_qubits[0].id
+        is_possible = (control, target) in self.connectivity
+        if not is_possible and (target, control) not in self.connectivity:
+            raise RuntimeError("The provided connectivity does not "
+                               "allow to execute the CNOT gate {}."
+                               .format(str(cmd)))
+        return not is_possible
+
+    def _send_cnot(self, cmd, control, target, flip=False):
+        def cmd_mod(command):
+            command.tags = deepcopy(cmd.tags) + command.tags
```

Comment:
The code below doesn't add any tags, so command.tags should always be empty. Correct?

Wouldn't the code look simpler if you just create the correct Command objects and send to the next engine?

Reply:
Fair enough